src/drivers/mmc/host/sdhci/sdhci_ops.c

```diff
@@ -0,0 +1,195 @@
+	log_debug("USDHC_AUTOCMD12_ERR_STATUS =0x%08x", sdhci_readl(host, 0x3C));
+	log_debug("USDHC_HOST_CTRL_CAP        =0x%08x", sdhci_readl(host, 0x40));
+	log_debug("USDHC_WTMK_LVL             =0x%08x", sdhci_readl(host, 0x44));
+	log_debug("USDHC_MIX_CTRL             =0x%08x", sdhci_readl(host, 0x48));
+	log_debug("USDHC_FORCE_EVENT          =0x%08x", sdhci_readl(host, 0x50));
+	log_debug("USDHC_ADMA_ERR_STATUS      =0x%08x", sdhci_readl(host, 0x54));
+	log_debug("USDHC_ADMA_SYS_ADDR        =0x%08x", sdhci_readl(host, 0x58));
+	log_debug("USDHC_DLL_CTRl             =0x%08x", sdhci_readl(host, 0x60));
+	log_debug("USDHC_DLL_STATUS           =0x%08x", sdhci_readl(host, 0x64));
+	log_debug("USDHC_CLK_TUNE_CTRL_STATUS =0x%08x", sdhci_readl(host, 0x68));
+	log_debug("USDHC_VEND_SPEC            =0x%08x", sdhci_readl(host, 0xC0));
+	log_debug("USDHC_MMC_BOOT             =0x%08x", sdhci_readl(host, 0xC4));
+	log_debug("USDHC_VEND_SPEC2           =0x%08x", sdhci_readl(host, 0xC8));
+}
+
+#define RETRIES    1000
+#define TIMEOUT    1000
+static void sdhci_cmd_done(struct sdhci_host *host, struct mmc_request *req) {
+	int i;
+
+	assert(host);
+	assert(req);
+
+	if (req->cmd.flags & MMC_RSP_136) {
+		/* CRC is stripped so we need to do some shifting. */
+		for (i = 0; i < 4; i++) {
+			req->cmd.resp[i] = sdhci_readl(host,
+					USDHC_CMD_RSP0 + (3 - i) * 4) << 8;
+			if (i != 3) {
+				req->cmd.resp[i] |= sdhci_readb(host,
+						USDHC_CMD_RSP0 + (3 - i) * 4 - 1);
+			}
+		}
+	} else {
+		req->cmd.resp[0] = sdhci_readl(host, USDHC_CMD_RSP0);
+	}
+}
+
+static void imx6_usdhc_send_cmd(struct sdhci_host *host, int cmd_idx, uint32_t arg, int flags) {
+	uint32_t wcmd = 0;
+	int retry = RETRIES;
+
+	wcmd |= (cmd_idx & 0x3f) << 24;
+
+	if (!(flags & MMC_RSP_PRESENT)) {
+	} else if (flags & MMC_RSP_136) {
+		wcmd |= 0x1 << 16;
+	} else if (flags & MMC_RSP_BUSY){
+		wcmd |= 0x3 << 16;
+	} else {
+		wcmd |= 0x2 << 16;
+	}
+
+	if (flags & MMC_RSP_CRC) {
+		wcmd |= 1 << 19;
+	}
+
+	if (flags & MMC_DATA_XFER) {
+		wcmd |= 0x1 << 21;
+		/* enable DMA in imx it places in MIX_CTRL */
+		wcmd |= 0x1;
+		if (flags & MMC_DATA_READ) {
+			wcmd |= 0x10;
+		}
+	}
+
+	sdhci_writel(host, USDHC_CMD_ARG, arg);
+
+	log_debug("send cmd: 0x%08x idx = %d arg = %x", wcmd, cmd_idx, arg);
+
+	sdhci_writel(host, USDHC_CMD_XFR_TYP, wcmd);
+
+	/* Wait command to be completed */
+	while(!(sdhci_readl(host, USDHC_INT_STATUS) & 0x1) && retry > 0) {
+		usleep(USEC_PER_MSEC);
+		retry--;
+	}
+
+	if (retry == 0) {
+		log_debug("Timeout!");
+	}
+
+	log_debug("USDHC_INT_STATUS=%08x", sdhci_readl(host, USDHC_INT_STATUS));
+	sdhci_writel(host, USDHC_INT_STATUS, 0x1);
+}
+
+static void sdhci_reset(struct sdhci_host *host) {
+	sdhci_writel(host, USDHC_SYS_CTRL, sdhci_readl(host, USDHC_SYS_CTRL) | USDHC_SYS_CTRL_RSTA);
+
+	int timeout = TIMEOUT;
+	while(timeout--);
+
+	if (sdhci_readl(host, USDHC_SYS_CTRL) & USDHC_SYS_CTRL_RSTA) {
+		log_error("Reset timeout");
```

Comment:
In case if reset is unsuccessful how is it possible to continue?

Reply:
You use this function in probe, which seems to be broken if this one failed

Reply:
Fixed